src/main/java/jenkins/plugins/logstash/LogstashWriter.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2014 Rusty Gerard and Liam Newman
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.plugins.logstash;
+
+
+import hudson.model.AbstractBuild;
+import jenkins.model.Jenkins;
+import jenkins.plugins.logstash.persistence.BuildData;
+import jenkins.plugins.logstash.persistence.IndexerDaoFactory;
+import jenkins.plugins.logstash.persistence.LogstashIndexerDao;
+import net.sf.json.JSONObject;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.exception.ExceptionUtils;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * A writer that wraps all Logstash DAOs.  Handles error reporting and per build connection state.
+ * Each call to write (one line or multiple lines) sends a Logstash payload to the DAO.
+ * If any write fails, writer will not attempt to send any further messages to logstash during this build.
+ *
+ * @author Rusty Gerard
+ * @author Liam Newman
+ * @since 1.0.5
+ */
+public class LogstashWriter {
+
+  final OutputStream errorStream;
+  final AbstractBuild build;
```

Comment:
Minor nitpick: it's a good idea to include the type arguments, though the Jenkins codebase is really bad about doing it.

`final AbstractBuild<?, ?> build;`